text/0031-corejam.md

```diff
@@ -0,0 +1,644 @@
+
+Existing functionality relied upon by parachains will continue to be provided as a special case under a more general and permissionless model which is detailed presently and known as *CoreJam*. Transition of Polkadot to be in line with the present proposal will necessarily imply some minor alterations of formats utilized by Cumulus, Smoldot and other light-client APIs (see the section on Compatibility). However, much of the underlying logic (in particular, consensus, disputes and availability) is retained, though its application is generalised. This proposal will only make note of the expectations regarding the changes, and presumes continuation of all other logic.
+
+As part of this model, we introduce a number of new and interrelated concepts: *Work Package*, *Work Class*, *Work Item*, *Work Output*, *Work Result*, *Work Report* and *Work Package Attestation*, *Work Class Trie*.
+
+Focussing on continuity and reuse of existing logic, it is unsurprising that many of these concepts already analogues in the Parachains model, albeit ones with a less general definition. While this mapping can be helpful to quickly create an approximate understanding of the new concepts for those already familiar with Polkadot, care must be taken not to inadvertantly make incorrect presumptions over exact details of their relationships, constraints, timing, provisions and APIs. Nonetheless, they are provided here for whatever help they may be.
+
+| CoreJam model                | Legacy model      | Context |
+| --- | --- | --- |
+| *Work Package*               | Proof-of-Validity | Untrusted data provided to RcVG |
+| *Work Item*                  | Proof-of-Validity | State-transition inputs and witness |
+| *Work Output*                | Candidate         | State-transition consequence |
+| *Work Report*                | Candidate         | Target of Attestation |
+| *(Work Package) Attestation* | Attestation       | Output signed by members of RcVG |
+| *Reporting*                  | Attestation       | Placement of Attestation on-chain |
+| *Integration*                | Inclusion         | Irreversible transition of state |
+| *Builder*                    | Collator          | Creator of data worthy of Attestation |
+
+Additionally, the *Work Class Trie* has no immediate analogue, but may be considered as the Relay-chain state used to track the code and head data of the parachains.
+
+### Overview
+
+```rust
+mod v0 {
+    const PROGRESS_WEIGHT_PER_PACKAGE: Weight = MAX_BLOCK_WEIGHT * 3 / 4;
+    type WorkClass = u32;
+    type WorkPayload = Vec<u8>;
+    struct WorkItem {
+        class: WorkClass,
+        payload: WorkPayload,
+    }
+    type MaxWorkItemsInPackage = ConstU32<4>;
+    type MaxWorkPackagePrerequisites = ConstU32<4>;
```

Comment:
```suggestion
    type MaxWorkPackagePrerequisites = ConstU32<4>;
```

From my understanding isn't the number of prerequisites limited to one? Also, I don't see where this is being used.

Reply:
Old code - should be removed.